### PR TITLE
Add upgrade::map_with_addr

### PR DESCRIPTION
**Pull request by tomaka**
Similar to `upgrade::map`, but resolves the address and provides it in the closure.

### CODE REVIEW

File: core/src/upgrade/map_addr.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+use std::io::Error as IoError;
+use futures::{future, prelude::*};
+use tokio_io::{AsyncRead, AsyncWrite};
+use upgrade::{ConnectionUpgrade, Endpoint};
+use Multiaddr;
+
+/// Applies a closure on the output of a connection upgrade.
+#[inline]
+pub fn map_with_addr<U, F, I, O>(upgrade: U, map: F) -> MapAddr<U, F>
+    where F: FnOnce(I, &Multiaddr) -> O
```
Comment (twittner) [inline core/src/upgrade/map_addr.rs]: `map` only constraints the type parameters in its `ConnectionUpgrade` impl, so for consistency we should do the same here, no?

Comment (tomaka) [reply core/src/upgrade/map_addr.rs]: In practice I was getting inference errors for the `Multiaddr`, so I added the constraint.
I don't know what we're supposed to do in theory.

**Review by twittner — APPROVED** [review]

File: core/src/upgrade/map_addr.rs (hunk 1 above)
Comment (twittner) [reply core/src/upgrade/map_addr.rs]: Maybe it is worth revisiting `map` at some point and do the same there, just for consistency. I do not have a strong preference, but if one approach causes us problems I am all for the other one ;-)

Comment (tomaka) [reply core/src/upgrade/map_addr.rs]: `map` doesn't have the `Multiaddr` parameter, so no inference error.

**Pull request merged**